consensus/istanbul/core/roundchange.go

```diff
@@ -177,67 +179,97 @@ func (c *core) handleRoundChange(msg *istanbul.Message) error {
 
 	roundView := rc.View
 
-	// Add the ROUND CHANGE message to its message set and return how many
-	// messages we've got with the same round number and sequence number.
-	num, err := c.roundChangeSet.Add(roundView.Round, msg)
-	if err != nil {
-		logger.Warn("Failed to add round change message", "message", msg, "err", err)
+	// Add the ROUND CHANGE message to its message set.
+	if err := c.roundChangeSet.Add(roundView.Round, msg); err != nil {
+		logger.Warn("Failed to add round change message", "roundView", roundView, "err", err)
 		return err
 	}
-	logger.Trace("Got round change message", "num", num, "message_round", roundView.Round)
+	logger.Trace("Got round change message", "message_round", roundView.Round)
+
+	// Skip to the highest round we know one honest validator is at, but
+	// don't start a round until we a quorum who want to start a given round.
+	ffRound := c.roundChangeSet.MaxRound(c.valSet.F() + 1)
+	quorumRound := c.roundChangeSet.MaxOnOneRound(c.valSet.MinQuorumSize())
```

Comment:
Because I was able to convince myself that starting a new round with round change messages for different rounds could result in a correctness violation. I think that this is still live because a quorum of honest validators will eventually converge on one round through timeouts and `ffround`.

This is the trickiest part of this PR.